core/data_transfer/impl/graphsync/graphsync_receiver.cpp

```diff
@@ -0,0 +1,174 @@
+            true,
+            sender,
+            channel_state->channel.base_cid,
+            channel_state->channel.selector->getRawBytes()));
+        event.code = EventCode::PROGRESS;
+        notifySubscribers(event, *channel_state);
+      }
+    }
+    return outcome::success();
+  }
+
+  void GraphsyncReceiver::receiveError() {
+    logger_->warn("Receive Error");
+  }
+
+  outcome::result<void> GraphsyncReceiver::sendResponse(
+      const PeerId &peer, bool is_accepted, const TransferId &transfer_id) {
+    DataTransferMessage response = createResponse(is_accepted, transfer_id);
+    OUTCOME_TRY(sender, network_->newMessageSender(peer));
+    OUTCOME_TRY(sender->sendMessage(response));
+    return outcome::success();
+  }
+
+  void GraphsyncReceiver::subscribeToEvents(
+      const std::shared_ptr<Subscriber> &subscriber) {
+    subscribers_.push_back(subscriber);
+  }
+
+  void GraphsyncReceiver::unsubscribe(
+      const std::shared_ptr<Subscriber> &subscriber) {
+    subscribers_.erase(
+        std::remove(subscribers_.begin(), subscribers_.end(), subscriber),
+        subscribers_.end());
+  }
+
+  outcome::result<void> GraphsyncReceiver::sendGraphSyncRequest(
+      const PeerId &initiator,
+      const TransferId &transfer_id,
+      bool is_pull,
+      const PeerId &sender,
+      const CID &root,
+      gsl::span<const uint8_t> selector) {
+    ExtensionDataTransferData extension_data{.transfer_id = transfer_id,
+                                             .initiator = initiator.toBase58(),
+                                             .is_pull = is_pull};
+    OUTCOME_TRY(extension, encodeDataTransferExtension(extension_data));
+
+    graphsync_->makeRequest(
+        sender,
+        boost::none,
+        root,
+        selector,
+        {extension},
+        [this, initiator, transfer_id, sender](
+            ResponseStatusCode code, std::vector<Extension> extensions) {
+          Event event{.code = EventCode::ERROR,
+                      .message = "",
+                      .timestamp = clock::UTCClockImpl().nowUTC()};
+
+          ChannelId channel_id{.initiator = initiator, .id = transfer_id};
+          auto channel = this->graphsync_manager_->getChannelByIdAndSender(
+              channel_id, sender);
+          if (!channel) {
+            event.code = EventCode::ERROR;
+            event.message = "cannot find a matching channel for this request";
+          } else if (isError(code)) {
+            event.code = EventCode::ERROR;
+            event.message = statusCodeToString(code);
+          } else if (isSuccess(code)) {
+            event.code = EventCode::COMPLETE;
+          }
+
+          this->notifySubscribers(event, *channel);
+        });
+    return outcome::success();
+  }
+
+  void GraphsyncReceiver::notifySubscribers(const Event &event,
+                                            const ChannelState &channel_state) {
+    for (auto subscriber : subscribers_) {
```

Comment:
`for (auto &subscriber : subscribers_) {`
In controlled circumstances pass shared_ptr by reference is better, than make a copy